list.py

```diff
@@ -0,0 +1,248 @@
+        self.find_node(self.index_transform(index)).elem = elem
+        
+
+    def __iter__(self):
+        return ListIterator(self.head)
+
+
+    def index_transform(self, index: int) -> int:
+        '''
+        Transform negative indices to correct positive value
+        '''
+        return index + self.length if index < 0 else index
+
+
+    def check_index_borders(self, index, border = 1):
+        '''
+        check index in: 
+            [-len; len): border = 1 (default)
+            [-len; len]: border = 0
+        '''
+        if index > self.length - border or index < -self.length:
+            self.raise_exception_oob(index)
+
+
+    def raise_exception_oob(self, index):
+        raise Exception(f"Index {index} is out of bound! List length: {self.length}")
+
+    
+    def find_node(self, index) -> Node:
+        '''
+        Find node by index
+        '''
+        curr_node = self.head
+
+        count = 0
+
+        while count < index:
+            curr_node = curr_node.next
+            count += 1
+
+        return curr_node
+
+
+    # C
+    def push(self, elem):
+        '''
+        Add element to tail list
+        '''
+        self.insert(self.length, elem)
+
+
+    # C
+    def insert(self, index, elem):
+        '''
+        Insert element in list by moving to right all element standing on the right
+        '''
+        self.check_index_borders(index, border = 0)
+
+        # empty list
+        if self.head == None:
+            self.head = self.tail = Node(elem)
+            self.length += 1 # length
+            return
+
+        index = self.index_transform(index)
+
+        # insert first element
+        if index == 0:
+            self.head = Node(elem, self.head)
+            self.length += 1 # length
+            return
+
+        # insert in tail
+        if index == self.length: # tail
+            self.tail.next = Node(elem)
+            self.tail = self.tail.next
+            self.length += 1 # length
+            return
+
+        prev_node = self.find_node(index - 1)
+        prev_node.next = Node(elem, prev_node.next)
+        self.length += 1
+
+
+    # D
+    def remove(self, index):
+        '''
+        Remove element by index
+        '''
+        if self.check_index_borders(index, border = 1):
+            self.raise_exception_oob(index)
+
+        index = self.index_transform(index)
+
+        if index == 0:
+            self.head = self.head.next
+            self.length -= 1
+            return
+
+        prev_node = self.find_node(index - 1)
+
+        if index == self.length - 1:
+            self.tail = prev_node
+        
+        prev_node.next = prev_node.next.next
+        self.length -= 1
+
+
+    def length(self) -> int:
+        return self.length
+
+
+    def find(self, elem) -> int | None:
+        '''
+        Try to find element in list
+        '''
+        curr_node = self.head
+
+        count = 0
+
+        while curr_node: # check for None
```

Comment:
This could be rewritten to a for loop, because this class is has `__iter__` method. Or we could even use iterator tools to do that